### PR TITLE
Conformitron with cost optimizations

### PR DESCRIPTION
**Pull request by Howlla**
*Issue #, if available:*

*Description of changes:* Code for conformitron mult-cluster deployment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cdk.json [hunk 1]
```diff
@@ -1,3 +1,38 @@
 {
-  "app": "npx ts-node dist/lib/common/default-main.js"
+  "app": "npx ts-node dist/lib/common/default-main.js",
+  "context":{
+    "conformitron":{
+      "amp" : {
+        "endpoint":"https://aps-workspaces.us-west-2.amazonaws.com/workspaces/ws-b08fda60-7e79-450c-972d-262ebac98c3e/",
```
Comment (elamaran11) [inline cdk.json:6]: Please remove this hard code values from your `cdk.json` file. We should populate it via context file via env variables which should be documented in your doc like backstage pattern document.

File: cdk.json [hunk 2]
```diff
@@ -1,3 +1,38 @@
...
+        "arn":"arn:aws:aps:us-west-2:867286930927:workspace/ws-b08fda60-7e79-450c-972d-262ebac98c3e"
+      },
+      "amg:": {
+        "endpoint":"https://g-76edcf29d5.grafana-workspace.us-west-2.amazonaws.com"
```
Comment (elamaran11) [inline cdk.json:10]: Likewise this should also be removed

File: docs/patterns/multi-cluster-conformitron.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
+# Multi Cluster Conformitron Testing
+
```
Comment (elamaran11) [inline docs/patterns/multi-cluster-conformitron.md:2]: Lets start with a little bit of background. You can talk about what is conformitron, what is the need for the same, what are we accomplishing etc etc. YOu can also link our conformitron blog.

File: docs/patterns/multi-cluster-conformitron.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+## Objective
+
+1. Creating a pipeline to deploy 8 EKS cluster in a region, with a Continuous Deployment  triggered upon a commit to the repository that holds the pipeline configuration.
+
+1. Configure the clusters with different architectures (x86 and arm) and different Kuberentes versions and use fluxCD pipeline to deploy all the available [EKS Anywhere Addons](https://github.com/aws-samples/eks-anywhere-addons).
```
Comment (elamaran11) [inline docs/patterns/multi-cluster-conformitron.md:7]: And Bottlerocket.

File: docs/patterns/multi-cluster-conformitron.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+## Deploying
+
+Once all pre-requisites are set you are ready to deploy the pipeline. Run the following command from the root of this repository to deploy the pipeline stack:
+
+```bash
+make pattern multi-cluster-conformitron deploy multi-cluster-central-pipeline
+```
+
+Now you can go to [AWS CodePipeline console](https://eu-west-1.console.aws.amazon.com/codesuite/codepipeline/pipelines), and see how it was automatically created to deploy multiple Amazon EKS clusters to different environments.
```
Comment (elamaran11) [inline docs/patterns/multi-cluster-conformitron.md:64]: What about deleting instructions, screenshots of working environment and documentation showing how to run the optimization SSM doc setup.

File: lib/multi-cluster-construct/CostOptimizationSSM/README.md [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# SSM Cost Optimizations for conformitron clusters
```
Comment (elamaran11) [inline lib/multi-cluster-construct/CostOptimizationSSM/README.md:1]: Could we remove this and this to main documentation of this pattern. lets not have two disconnected docs.

File: lib/multi-cluster-construct/CostOptimizationSSM/README.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+Running 9 testing clusters for 24 hours results in a daily spend of $300+
+
+To minimize these costs we have written a systems manager automation which automatically scales down autoscaling group to 0 desired nodes during off-business hours.
+
+On weekdays 5 PM PST clusters are scaled to 0 -> CRON EXPRESSION:  0 17 ? * MON-FRI *
+On weekdays 5 AM PST clusters are scaled to 1 -> CRON EXPRESSION:  0 05 ? * MON-FRI *
```
Comment (elamaran11) [inline lib/multi-cluster-construct/CostOptimizationSSM/README.md:8]: Lets also show CLI based instructions on how to set it up

File: lib/multi-cluster-construct/CostOptimizationSSM/ScaleUpEKStoOne.YAML [hunk 8]
```diff
@@ -0,0 +1,109 @@
+schemaVersion: '0.3'
```
Comment (elamaran11) [inline lib/multi-cluster-construct/CostOptimizationSSM/ScaleUpEKStoOne.YAML:1]: Could we use some env var variables to populate cluster names etc etc so its not direct hard code to the yaml. Also why is the yaml file `YAML` format?

**Review by elamaran11 — CHANGES_REQUESTED** [review]
@Howlla Great work, almost near to finish line. Have some feedback.
1. Documentation should be improved for completeness. Overview should be added, delete instructions, screenshots of working environment etc etc to be added.
2. Lets not have separate instructions for Cost Optimization documentation. Have this in single document. Think about someone want to replicate the entire setup using your documentation. Lets have complete setup instructions.
3. `cdk.json` should not have sensitive info.

Appreciate if you can fix this, once done, can review again.

**Comment by elamaran11** [comment]
@Howlla Could you please bring this PR to closure. Its long pending.

File: docs/patterns/multi-cluster-conformitron.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
+# Multi Cluster Conformitron Testing
```
Comment (shapirov103) [inline docs/patterns/multi-cluster-conformitron.md:1]: Let's call it `Multi-cluster pattern with observability cost optimizations and metrics aggregation`. Customers won't understand Conformitron but we can call it out in the intro.

File: lib/multi-cluster-construct/CostOptimizationSSM/README.md (hunk 6 above)
Comment (shapirov103) [reply lib/multi-cluster-construct/CostOptimizationSSM/README.md:1]: +1

Comment (shapirov103) [inline lib/multi-cluster-construct/CostOptimizationSSM/README.md:1]: Please remove the CostOptimizationSSM folder from lib

File: lib/multi-cluster-construct/eksa-secret-stores.ts [hunk 10]
```diff
@@ -0,0 +1,69 @@
+import * as blueprints from '@aws-quickstart/eks-blueprints';
+import * as eks from "aws-cdk-lib/aws-eks";
+import { Construct } from 'constructs';
+import { dependable } from '@aws-quickstart/eks-blueprints/dist/utils';
+
+export class EksAnywhereSecretsAddon implements blueprints.ClusterAddOn {
```
Comment (shapirov103) [inline lib/multi-cluster-construct/eksa-secret-stores.ts:6]: Let's rename this to something that is more semantically correct: Looks like it is an addon that sets up secret store for secret manager and paramstore. Let's call `ClusterSecretStoreAddon`.

Applies to the addon name and file name

File: lib/multi-cluster-construct/grafana-monitor-builder.ts [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+import { GrafanaOperatorSecretAddon } from './grafana-operator-secret-addon';
+import * as fs from 'fs';
+
+export class GrafanaMonitoringConstruct {
+
+    build(scope: Construct, id: string, contextAccount?: string, contextRegion?: string ) {
+
+        const stackId = `${id}-grafana-monitor`;
+
+        const account = contextAccount! || process.env.COA_ACCOUNT_ID! || process.env.CDK_DEFAULT_ACCOUNT!;
```
Comment (shapirov103) [inline lib/multi-cluster-construct/grafana-monitor-builder.ts:13]: What is this variable `COA_ACCOUNT_ID`? This maybe confusing to the end user of the pattern. Let's rename it to be something less confusing and document in the pattern doc.

Applies elsewhere.

File: lib/multi-cluster-construct/grafana-monitor-builder.ts [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+        const ampWorkspaceName = "conformitronWorkspace";
+        const ampEndpoint = blueprints.utils.valueFromContext(scope, "conformitron.amp.endpoint", "https://aps-workspaces.<region>.amazonaws.com/workspaces/<workspace-id>/");
+        const ampWorkspaceArn = blueprints.utils.valueFromContext(scope, "conformitron.amp.arn", "arn:aws:aps:<region>:<accountid>:workspace/<workspace-id>");
+        
+        const ampAddOnProps: blueprints.AmpAddOnProps = {
+            ampPrometheusEndpoint: ampEndpoint,
+            ampRules: {
+                ampWorkspaceArn: ampWorkspaceArn,
+                ruleFilePaths: [
+                    __dirname + '/../common/resources/amp-config/alerting-rules.yml',
```
Comment (shapirov103) [inline lib/multi-cluster-construct/grafana-monitor-builder.ts:35]: These config rules seem to be specific to the pattern, we could move them under multi-cluster-construct/resources. Let's discuss if needed.

File: lib/multi-cluster-construct/grafana-operator-secret-addon.ts [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+                            name: "ssm-parameter-store",
+                            kind: "ClusterSecretStore",
+                        },
+                        target: {
+                            name: "grafana-admin-credentials"
+                        },
+                        data: [
+                            {
+                                secretKey: "GF_SECURITY_ADMIN_APIKEY",
+                                remoteRef: {
```
Comment (shapirov103) [inline lib/multi-cluster-construct/grafana-operator-secret-addon.ts:63]: "/cdk-accelerator/grafana-api-key" - how will the customers know to set it up?  Also do we need to call out cdk-accelerator here?

File: lib/multi-cluster-construct/multi-cluster-builder.ts [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        
+        
+
+        return blueprints.ObservabilityBuilder.builder()
+            .account(accountID)
+            .region(awsRegion)
+            .withAmpProps(ampAddOnProps)
+            .enableOpenSourcePatternAddOns()
+            .addOns(
+                new blueprints.addons.FluxCDAddOn({
```
Comment (shapirov103) [inline lib/multi-cluster-construct/multi-cluster-builder.ts:49]: Let's call it in the docs that by default it sets up Flux CD with these GitHib repos, but that is where customers can use the GitOps tooling of their choice, as well as repos.

File: lib/multi-cluster-construct/pipeline.ts [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+            })
+            .build(scope, "multi-cluster-central-pipeline", {
+                env: {
+                    account: process.env.CDK_DEFAULT_ACCOUNT,
+                    region: region,
+                }
+            });
+    }
+
+    getClusterProps() {
```
Comment (shapirov103) [inline lib/multi-cluster-construct/pipeline.ts:131]: why do we need this method? It seems easier to just declare a const at the top like:

```ts
const defaultClusterProps : blueprints.MngClusterProviderProps = {
            maxSize : 2,
            minSize : 1,
            desiredSize: 1,
            diskSize: 100
        };
```

And use it where needed.

File: lib/multi-cluster-construct/pipeline.ts [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+        /* TODO: Seperate region for clusters than infra account region, 
+           trust policy is created when pipeline is bootstrapped.
+           It will be helpful for enterprise customers.
+           Similar to approach in multi-region-construct pattern
+        */
+
+        const clusterProps = this.getClusterProps();
+
+        for(const version of CLUSTER_VERSIONS) {
+            const blueprintBuilderX86 = new MultiClusterBuilderConstruct().create(scope, accountID, region);
```
Comment (shapirov103) [inline lib/multi-cluster-construct/pipeline.ts:42]: style: let's componentize a bit more. I suggest a method  `buildClusterProps(instanceTypes, amiType)` and call it for x86 and arm. That will reduce this method size and make it more readable.

**Review by shapirov103 — COMMENTED** [review]
@Howlla great work and much needed pattern. Please see some comments I left. It would be nice to test it out with 1.15.0 blueprints that we released last week.

File: cdk.json (hunk 1 above)
Comment (Howlla) [reply cdk.json]: removed from cdk.json and captured it via env variables instead

File: cdk.json (hunk 2 above)
Comment (Howlla) [reply cdk.json]: removed

File: docs/patterns/multi-cluster-conformitron.md (hunk 3 above)
Comment (Howlla) [reply docs/patterns/multi-cluster-conformitron.md:2]: Please give more feedback around the doc @shapirov103 @elamaran11  and I will make it work

File: docs/patterns/multi-cluster-conformitron.md (hunk 4 above)
Comment (Howlla) [reply docs/patterns/multi-cluster-conformitron.md]: mentioned

File: docs/patterns/multi-cluster-conformitron.md (hunk 5 above)
Comment (Howlla) [reply docs/patterns/multi-cluster-conformitron.md:111]: Pending still

File: lib/multi-cluster-construct/CostOptimizationSSM/README.md (hunk 6 above)
Comment (Howlla) [reply lib/multi-cluster-construct/CostOptimizationSSM/README.md]: Done

File: lib/multi-cluster-construct/CostOptimizationSSM/ScaleUpEKStoOne.YAML (hunk 8 above)
Comment (Howlla) [reply lib/multi-cluster-construct/CostOptimizationSSM/ScaleUpEKStoOne.YAML]: Pending still

File: docs/patterns/multi-cluster-conformitron.md (hunk 9 above)
Comment (Howlla) [reply docs/patterns/multi-cluster-conformitron.md]: renamed, open to more feedback on doc

File: lib/multi-cluster-construct/CostOptimizationSSM/README.md (hunk 6 above)
Comment (Howlla) [reply lib/multi-cluster-construct/CostOptimizationSSM/README.md]: i have moved it to a multi-cluster-construct/resources folder instead

File: lib/multi-cluster-construct/eksa-secret-stores.ts (hunk 10 above)
Comment (Howlla) [reply lib/multi-cluster-construct/eksa-secret-stores.ts]: Done

File: lib/multi-cluster-construct/grafana-monitor-builder.ts (hunk 11 above)
Comment (Howlla) [reply lib/multi-cluster-construct/grafana-monitor-builder.ts]: Fixed this variable name to ACCOUNT_ID

File: lib/multi-cluster-construct/grafana-monitor-builder.ts (hunk 12 above)
Comment (Howlla) [reply lib/multi-cluster-construct/grafana-monitor-builder.ts:35]: i have moved it to a multi-cluster-construct/resources folder instead

File: lib/multi-cluster-construct/grafana-operator-secret-addon.ts (hunk 13 above)
Comment (Howlla) [reply lib/multi-cluster-construct/grafana-operator-secret-addon.ts:63]: called out in documentation now

File: lib/multi-cluster-construct/multi-cluster-builder.ts (hunk 14 above)
Comment (Howlla) [reply lib/multi-cluster-construct/multi-cluster-builder.ts:49]: Mentioned in doc

File: lib/multi-cluster-construct/pipeline.ts (hunk 15 above)
Comment (Howlla) [reply lib/multi-cluster-construct/pipeline.ts:131]: refactored

File: lib/multi-cluster-construct/pipeline.ts (hunk 16 above)
Comment (Howlla) [reply lib/multi-cluster-construct/pipeline.ts:42]: refactored to use buildClusterProps function instead

**Comment by Howlla** [comment]
Pending architecture diagram still.

**Review by elamaran11 — COMMENTED** [review]
Overall, everything now LGTM. @shapirov103 Please share if all your comments are fixed and let me know. I can check once again and do merge.

File: docs/patterns/multi-cluster-conformitron.md (hunk 5 above)
Comment (shapirov103) [reply docs/patterns/multi-cluster-conformitron.md:172]: let's resolve this

**Review by shapirov103 — APPROVED** [review]
There are some minor things to improve on the doc side, like making sure you use dots at the end of the sentences, and proper casing, some spelling errors like seperate should be separate, vpc should be VPC, and such. I am willing to defer these, there is always space to improve. Please check some pending comments from Ela, like instructions to delete, which may cause some issue with ArgoCD resource being orphaned. We should be explicit about it, if there is some manual effort and gotchas, we can call them out.

**Review by elamaran11 — APPROVED** [review]
LGTM. @Howlla This is bar raising work, im good to merge this. Hope you can support the pattern in continuous basis for ongoing issues with Conformitron.

**Pull request merged**